File: sample/cpp/01_mandelbrot.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/std/time.h"
#include "generated/utils/png.h"

// 01: Sample that outputs the Mandelbrot set as a PNG image.
// Syntax is kept straightforward with future transpilation in mind.

int64 escape_count(float64 cx, float64 cy, int64 max_iter) {
    /* Return the iteration count until divergence for one point (cx, cy). */
    float64 x = 0.0;
    float64 y = 0.0;
    for (int64 i = 0; i < max_iter; ++i) {
        float64 x2 = x * x;
        float64 y2 = y * y;
        if (x2 + y2 > 4.0)
            return i;
        y = 2.0 * x * y + cy;
        x = x2 - y2 + cx;
    }
    return max_iter;
}

::std::tuple<int64, int64, int64> color_map(int64 iter_count, int64 max_iter) {
    /* Convert an iteration count to RGB. */
    if (iter_count >= max_iter)
        return ::std::make_tuple(0, 0, 0);
    float64 t = static_cast<float64>(iter_count) / static_cast<float64>(max_iter);
    int64 r = static_cast<int64>(255.0 * t * t);
    int64 g = static_cast<int64>(255.0 * t);
    int64 b = static_cast<int64>(255.0 * (1.0 - t));
    return ::std::make_tuple(r, g, b);
}

bytearray render_mandelbrot(int64 width, int64 height, int64 max_iter, float64 x_min, float64 x_max, float64 y_min, float64 y_max) {
    /* Generate RGB bytes for a Mandelbrot image. */
    bytearray pixels = bytearray{};
    float64 __hoisted_cast_1 = static_cast<float64>(height - 1);
    float64 __hoisted_cast_2 = static_cast<float64>(width - 1);
    float64 __hoisted_cast_3 = static_cast<float64>(max_iter);
    
    for (int64 y = 0; y < height; ++y) {
        float64 py = y_min + (y_max - y_min) * (static_cast<float64>(y) / __hoisted_cast_1);
        
        for (int64 x = 0; x < width; ++x) {
            float64 px = x_min + (x_max - x_min) * (static_cast<float64>(x) / __hoisted_cast_2);
            int64 it = escape_count(px, py, max_iter);
            int64 r;
            int64 g;
            int64 b;
            if (it >= max_iter) {
                r = 0;
                g = 0;
                b = 0;
            } else {
                float64 t = static_cast<float64>(it) / __hoisted_cast_3;
                r = static_cast<int64>(255.0 * t * t);
                g = static_cast<int64>(255.0 * t);
                b = static_cast<int64>(255.0 * (1.0 - t));
            }
            pixels.append(static_cast<uint8>(static_cast<int64>(r)));
            pixels.append(static_cast<uint8>(static_cast<int64>(g)));
            pixels.append(static_cast<uint8>(static_cast<int64>(b)));
        }
    }
    return pixels;
}

void run_mandelbrot() {
    int64 width = 1600;
    int64 height = 1200;
    int64 max_iter = 1000;
    str out_path = "sample/out/01_mandelbrot.png";
    
    float64 start = pytra::std::time::perf_counter();
    
    bytearray pixels = render_mandelbrot(width, height, max_iter, -(2.2), 1.0, -(1.2), 1.2);
    pytra::utils::png::write_rgb_png(out_path, width, height, pixels);
    
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("size:", width, "x", height);
    py_print("max_iter:", max_iter);
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_mandelbrot();
    return 0;
}

File: sample/py/01_mandelbrot.py
# 01: マンデルブロ集合を PNG 画像として出力するサンプルです。
# 将来のトランスパイルを意識して、構文はなるべく素直に書いています。

from time import perf_counter
from pylib import png


def escape_count(cx: float, cy: float, max_iter: int) -> int:
    """1点 (cx, cy) の発散までの反復回数を返す。"""
    x: float = 0.0
    y: float = 0.0
    for i in range(max_iter):
        x2: float = x * x
        y2: float = y * y
        if x2 + y2 > 4.0:
            return i
        y = 2.0 * x * y + cy
        x = x2 - y2 + cx
    return max_iter


def color_map(iter_count: int, max_iter: int) -> tuple[int, int, int]:
    """反復回数を RGB に変換する。"""
    if iter_count >= max_iter:
        return (0, 0, 0)

    # 簡単なグラデーション（青系 -> 黄系）
    t: float = iter_count / max_iter
    r: int = int(255.0 * (t * t))
    g: int = int(255.0 * t)
    b: int = int(255.0 * (1.0 - t))
    return (r, g, b)


def render_mandelbrot(
    width: int,
    height: int,
    max_iter: int,
    x_min: float,
    x_max: float,
    y_min: float,
    y_max: float,
) -> bytearray:
    """マンデルブロ画像の RGB バイト列を生成する。"""
    pixels: bytearray = bytearray()

    for y in range(height):
        py: float = y_min + (y_max - y_min) * (y / (height - 1))

        for x in range(width):
            px: float = x_min + (x_max - x_min) * (x / (width - 1))
            it: int = escape_count(px, py, max_iter)
            r: int
            g: int
            b: int
            if it >= max_iter:
                r = 0
                g = 0
                b = 0
            else:
                t: float = it / max_iter
                r = int(255.0 * (t * t))
                g = int(255.0 * t)
                b = int(255.0 * (1.0 - t))
            pixels.append(r)
            pixels.append(g)
            pixels.append(b)

    return pixels


def run_mandelbrot() -> None:
    width: int = 1600
    height: int = 1200
    max_iter: int = 1000
    out_path: str = "sample/out/mandelbrot_01.png"

    start: float = perf_counter()

    pixels: bytearray = render_mandelbrot(
        width,
        height,
        max_iter,
        -2.2,
        1.0,
        -1.2,
        1.2,
    )
    png.write_rgb_png(out_path, width, height, pixels)

    elapsed: float = perf_counter() - start
    print("output:", out_path)
    print("size:", width, "x", height)
    print("max_iter:", max_iter)
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_mandelbrot()
